pkg/scheduler/plugins/tdm/tdm.go

```diff
@@ -0,0 +1,333 @@
+		score := 0.0
+
+		if node.RevocableZone == "" {
+			return score, nil
+		}
+
+		if err := bp.availableRevocableZone(node.RevocableZone); err != nil {
+			klog.V(4).Infof("TDM not available %s", err)
+			return score, err
+		}
+
+		if !task.Preemptable {
+			klog.V(4).Infof("TDM task %s/%s is not preemptable task, skip to schedule to node %s", task.Namespace, task.Name, node.Name)
+			return score, nil
+		}
+
+		score = float64(v1alpha1.MaxNodeScore)
+
+		klog.V(4).Infof("TDM score for Task %s/%s on node %s is: %v", task.Namespace, task.Name, node.Name, score)
+		return score, nil
+	}
+
+	preemptableFn := func(preemptor *api.TaskInfo, preemptees []*api.TaskInfo) []*api.TaskInfo {
+		if preemptor.Preemptable {
+			klog.V(4).Infof("TDM task %s/%s is preemptable, do nothing skip", preemptor.Namespace, preemptor.Name)
+			return preemptees
+		}
+
+		var victims []*api.TaskInfo
+
+		// find preemptable tasks which appear on none revocable node
+		for _, task := range preemptees {
+			if !task.Preemptable || task.Status != api.Running {
+				continue
+			}
+
+			node, ok := ssn.Nodes[task.NodeName]
+			if !ok {
+				continue
+			}
+
+			if node.RevocableZone != "" {
+				continue
+			}
+
+			victims = append(victims, task)
+		}
+
+		active := map[string]struct{}{}
+
+		for rz := range bp.revocableZone {
+			if err := bp.availableRevocableZone(rz); err == nil {
+				active[rz] = struct{}{}
+			}
+		}
+
+		// if there are some active rz, then evict preemptable tasks from all the none revocable node
+		if len(active) > 0 {
+			victims = append(victims, bp.noneRevocableNodePreemptableTask(ssn)...)
+		}
+
+		victims = bp.maxVictims(victims)
+		klog.V(4).Infof("TDM victims are %+v", victims)
+
+		return victims
+	}
+
+	victimsFn := func() []*api.TaskInfo {
+		if lastEvictAt.Add(bp.evictPeriod).After(time.Now()) {
+			klog.V(4).Infof("TDM next evict time at %v", lastEvictAt)
+			return nil
+		}
+
+		klog.V(4).Infof("TDM start to find victims")
+
+		// find preemptable task on timeout revocable zone node
+		victims := make([]*api.TaskInfo, 0)
+		for rz := range bp.revocableZone {
+			if err := bp.availableRevocableZone(rz); err != nil {
+				klog.V(4).Infof("TDM revocable zone %v disactive, %v", rz, err)
+				// rz disactive, then evict preemptable tasks from the revocable node
+				victims = append(victims, bp.revocableNodePreemptableTask(rz, ssn)...)
+			}
+		}
+
+		victims = bp.maxVictims(victims)
+		// need to consider concurrency?
+		lastEvictAt = time.Now()
+
+		klog.V(4).Infof("TDM got %v victims", len(victims))
+
+		return victims
+	}
+
+	ssn.AddPredicateFn(bp.Name(), predicateFn)
+	ssn.AddNodeOrderFn(bp.Name(), nodeOrderFn)
+	ssn.AddPreemptableFn(bp.Name(), preemptableFn)
+	ssn.AddVictimTasksFns(bp.Name(), victimsFn)
+}
+
+func (bp *tdmPlugin) maxVictims(victims []*api.TaskInfo) []*api.TaskInfo {
+	targetNum := 0
+	switch bp.evictMaxNum.Type {
+	case intstr.Int:
+		targetNum = int(math.Min(float64(bp.evictMaxNum.IntValue()), float64(len(victims))))
```

Comment:
are we going to select Task/Pod by Job/PodGroup, e.g. not only get pod from one PodGroup.

Reply:
OK，I will use a new PR to handle this comments, after we support evictMaxNum params in workload.